src/vcTiffConvert.cpp

```diff
@@ -10,157 +10,592 @@
+        uint32_t green = vcTiff_To8BitSample(data.t_double[1]);
+        uint32_t blue = vcTiff_To8BitSample(data.t_double[2]);
+        uint32_t alpha = vcTiff_To8BitSample(data.t_double[3]);
+        *pColour = (red) | (green << 8) | (blue << 16) | (alpha << 24);
+      }
+      else
+      {
+        UD_ERROR_SET(udE_InvalidConfiguration);
+      }
+      break;
+    }
   }
 
-protected:
+  result = udE_Success;
+epilogue:
+  return result;
+}
 
-  virtual udError OnInit() = 0;
-  virtual void OnShutDown() = 0;
+static udError vcTiff_DecodeSample(void * pBuf, uint64_t pixelIndex, vcTiffImageFormat const & format, uint32_t sampleIndex, vcTiffSampleData *pSample)
+{
+  udError result;
 
-  TIFF *m_pTiff;
-  double m_pointResolution;
-  udDouble3 m_origin;
-  udConvertCustomItemFlags m_convertFlags;
-};
+  if (pBuf == nullptr || pSample == nullptr)
+    UD_ERROR_SET(udE_InvalidParameter);
 
-// TODO add TiffScanLineReader, TiffStripReader, TiffTIleReader.
-// And perhaps within these, objects to read the different types of data formats.
+  if (format.bitsPerSample == 1)
+  {
+    // A 1-bit int or float makes no sense
+    UD_ERROR_IF(format.sampleFormat != SAMPLEFORMAT_UINT, udE_InvalidConfiguration);
+    uint64_t bitIndex = pixelIndex *format.samplesPerPixel + sampleIndex;
+    uint8_t byte = ((uint8_t *)pBuf)[bitIndex / 8];
+    uint64_t value = uint64_t(byte & (1 << (bitIndex % 8))) == 0 ? 0 : 1;
+    pSample->t_uint64[sampleIndex] = value;
+  }
+  else if (format.bitsPerSample == 2)
+  {
+    // Currently samples less than 8 bits must be uint
+    UD_ERROR_IF(format.sampleFormat != SAMPLEFORMAT_UINT, udE_InvalidConfiguration);
+    uint64_t bitIndex = pixelIndex * format.samplesPerPixel + sampleIndex;
+    uint8_t byte = ((uint8_t *)pBuf)[bitIndex / 4];
+    uint32_t shft = (bitIndex % 4) * 2;
+    uint64_t value = uint64_t((byte >> shft) & 0x3);
+    pSample->t_uint64[sampleIndex] = value;
+  }
+  else if (format.bitsPerSample == 4)
+  {
+    // Currently samples less than 8 bits must be uint
+    UD_ERROR_IF(format.sampleFormat != SAMPLEFORMAT_UINT, udE_InvalidConfiguration);
+    uint64_t bitIndex = pixelIndex * format.samplesPerPixel + sampleIndex;
+    uint8_t byte = ((uint8_t *)pBuf)[bitIndex / 2];
+    uint32_t shft = (bitIndex % 2) * 4;
+    uint64_t value = uint64_t((byte >> shft) & 0xF);
+    pSample->t_uint64[sampleIndex] = value;
+  }
+  else if (format.bitsPerSample == 8)
+  {
+    // Currently do not support 8-bit floats
+    uint64_t ind = pixelIndex * format.samplesPerPixel + sampleIndex;
+    if (format.sampleFormat == SAMPLEFORMAT_INT)
+      pSample->t_int64[sampleIndex] = ((int8_t *)(pBuf))[ind];
+    else if (format.sampleFormat == SAMPLEFORMAT_UINT)
+      pSample->t_uint64[sampleIndex] = ((uint8_t *)(pBuf))[ind];
+    else
+      UD_ERROR_SET(udE_InvalidConfiguration);
+  }
+  else if (format.bitsPerSample == 16)
+  {
+    // Currently do not support 16-bit floats
```

Comment:
Just an FYI, I believe there's a function that can pack and/or unpack 16bit floats in udStream already. This is used for the 16bit float on the graphics card - I'm confident it won't be immediately usable for you though, so not a required change for this PR.